<a href="https://colab.research.google.com/github/rohansiddam/Python-Journey/blob/main/014%20-%20Project%2014%20(Loops%20%26%20Pandas%20DataFrames).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Instructions

#### Goal of the Project

This project is designed for you to practice and solve the activities that are based on the concepts covered in the following lessons:

1. While Loop, Data-Type Conversion and Conditional Statements.

2. Pandas DataFrame.


---

#### Getting Started:

1. Click on this link to open the Colab file for this project.

  https://colab.research.google.com/drive/1GO3d4EFI1qq70Ae-fDZ89243k3lVn-tn

2. Create a duplicate copy of the Colab file as described below.

  - Click on the **File menu**. A new drop-down list will appear.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/0_file_menu.png' width=500>

  - Click on the **Save a copy in Drive** option. A duplicate copy will get created. It will open up in the new tab on your web browser.

  <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/1_create_colab_duplicate_copy.png' width=500>

3. After creating the duplicate copy of the notebook, please rename it in the **YYYY-MM-DD_StudentName_Project14** format.

4. Now, write your code in the prescribed code cells.


---

### Activities


  

#### Activity 1: Create Pandas DataFrame

Create a pandas DataFrame by using the below link which has the dataset of NBA players in the CSV format:

**Dataset** : https://media.geeksforgeeks.org/wp-content/uploads/nba.csv

Also, print the first five rows of the dataset.

In [ ]:
# Create a pandas DataFrame
import pandas as pd
nba = pd.read_csv('https://media.geeksforgeeks.org/wp-content/uploads/nba.csv')
nba.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

---

#### Activity 1.2 : Check for Null values

Check for missing values in the DataFrame.

In [ ]:
# Check for missing values in the above Pandas series.
nba.isnull().sum()

Name         1
Team         1
Number       1
Position     1
Age          1
Height       1
Weight       1
College     85
Salary      12
dtype: int64

**Q:** Are there any missing values in the DataFrame ?

**A:** Yes there are.

---

#### Activity 1.3 : Get the Number of rows and columns

Get the number of rows and columns of the DataFrame.

In [ ]:
# Get the Number of rows and columns
nba.shape

(458, 9)

**Q:** How many rows are there in the DataFrame?

**A:** 458

---

#### Activity 1.4: Create Pandas series for the player 'Avery Bradley'

Using the `iloc[]` function, create a Pandas series for the player Avery Bradley and store it in a variable called `avery_bradley`.

In [ ]:
# Create pandas series for the player 'Avery Bradley'
avery_bradley = nba.iloc[0,:]
print(avery_bradley)

Name         Avery Bradley
Team        Boston Celtics
Number                 0.0
Position                PG
Age                   25.0
Height                 6-2
Weight               180.0
College              Texas
Salary           7730337.0
Name: 0, dtype: object


**Q**: Which team does Avery Bradely plays for?

**A**: Boston Celtics

---

#### Activity 2: Find the number of solutions for a Quadratic equation

A quadratic equation $ax² + b x + c = 0$ has either 0, 1, or 2 distinct solutions for real values of $x$. Given $a$, $b$ and $c$, you should only return the number of solutions that we can get from the equation.

The above quadratic equation is solved using the formula, \begin{array}{*{20}c} {x = \frac{{ - b \pm \sqrt {b^2 - 4ac} }}{{2a}}} & {{\rm{when}}} & {ax^2 + bx + c = 0} \\ \end{array}

The `±` in the above formula means there are **TWO** answers.

Do you see $b^2 − 4ac$ in the formula above? It is called the **Discriminant**, because it can "discriminate" between the possible types of answer:

  - when $b^2 − 4ac$ is **positive**, we get **two** Real solutions.

  - when $b^2 − 4ac$ is **zero**, we get just **one** real solution.

  - when $b^2 − 4ac$ is **negative**, we get a pair of **complex solution**.


Examples:
```
1. solutions(1, 0, -1) ➞ 2
# x² - 1 = 0 has two solutions (x = 1 and x = -1).

2. solutions(1, 0, 1) ➞ 0
# x² + 1 = 0 has no solutions.
```

**Note**: You do not need to calculate the solutions, just return how many solutions are there by calculating the value of $b^2 − 4ac$.


In [ ]:
# Number of solutions for a quadratic equation
a = int(input("Enter in a number: "))
b = int(input("Enter in a number: "))
c = int(input("Enter in a number: "))

if ((b ** b) - (4 * a * c)) > 0:
  print("There are 2 real solutions.")
if ((b ** b) - (4 * a * c)) == 0:
  print("There is one real solution.")
if ((b ** b) - (4 * a * c)) < 0:
  print("There is one pair of complex solutions.")

Enter in a number: 3
Enter in a number: 5
Enter in a number: 6
There are 2 real solutions.


---

### Submitting the Project:

1. After finishing the project, click on the **Share** button on the top right corner of the notebook. A new dialog box will appear.

  <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/2_share_button.png' width=500>

2. In the dialog box, make sure that '**Anyone on the Internet with this link can view**' option is selected and then click on the **Copy link** button.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/3_copy_link.png' width=500>

3. The link of the duplicate copy (named as **YYYY-MM-DD_StudentName_Project14**) of the notebook will get copied

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/4_copy_link_confirmation.png' width=500>

4. Go to your dashboard and click on the **My Projects** option.
   
   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/5_student_dashboard.png' width=800>

  <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/6_my_projects.png' width=800>

5. Click on the **View Project** button for the project you want to submit.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/7_view_project.png' width=800>

6. Click on the **Submit Project Here** button.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/8_submit_project.png' width=800>

7. Paste the link to the project file named as **YYYY-MM-DD_StudentName_Project14** in the URL box and then click on the **Submit** button.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/9_enter_project_url.png' width=800>